set window title on loading a repo

birgersp opened the pull request:
fix #1691

jesseduffield commented on pkg/gui/gui.go:
```diff
@@ -778,3 +784,18 @@ func (gui *Gui) OnUIThread(f func() error) {
 	})
 }
+
+func (gui *Gui) UpdateWindowTitle() error {
```
Could we put this function in `pkg/commands/oscommands/os.go`? We can then invoke it as `gui.OSCommand.UpdateWindowTitle()`.

birgersp replied:
Absolutely, I wasn't at all confident about that location

jesseduffield commented on pkg/gui/gui.go:
```diff
@@ -778,3 +784,18 @@ func (gui *Gui) OnUIThread(f func() error) {
+	if getWdErr != nil {
+		return getWdErr
+	}
+	cmd := exec.Command("cmd", "/C", "title", filepath.Base(path))
```
I wonder if we should be saying 'RepoName - Lazygit' or 'Lazygit - RepoName' to make it clearer. Happy for you to make the call on this.

Also we'll need to ensure that if our repo has spaces in the name that this command still works as expected.

jesseduffield replied:
Another thing: we have an existing approach for running something in a shell that's platform-agnostic. That is, we could go:
```
func (c *OSCommand) SetWindowTitle(title string) error {
	return c.Cmd.NewShell("title " + c.Quote(title)).Run()
}
```
But I'm not on windows and don't know whether that works off-hand

birgersp replied:
> I wonder if we should be saying 'RepoName - Lazygit' or 'Lazygit - RepoName' to make it clearer. Happy for you to make the call on this.
> 
> Also we'll need to ensure that if our repo has spaces in the name that this command still works as expected.

From the other windows I have open it seems a bit mixed, but Google Chrome and VSCode both have the "[custom title] - [application name]" format so I think maybe that makes most sense.

birgersp replied:
> Another thing: we have an existing approach for running something in a shell that's platform-agnostic. That is, we could go:
> 
> ```
> func (c *OSCommand) SetWindowTitle(title string) error {
> 	return c.Cmd.NewShell("title " + c.Quote(title)).Run()
> }
> ```
> 
> But I'm not on windows and don't know whether that works off-hand

Makes sense, but how would I invoke it from [gui.loadNewRepo](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/gui.go#L677)?

birgersp replied:
Nevermind, I got it :)

> Also we'll need to ensure that if our repo has spaces in the name that this command still works as expected.

Good thinking. Turns out that we need to write out the title without quotation marks, even when there are spaces.

So the command is like this: `title This is my window title`.

![image](https://user-images.githubusercontent.com/5260237/150121856-1c275cbb-1f4b-4a93-88b5-008e3b47f498.png)

jesseduffield reviewed (changes requested):
Nice work! Left a couple comments

birgersp commented:
@jesseduffield Idk if I did this PR process correctly (I probably shouldn't have amended and force pushed when doing changes, I'll keep that in mind going forward). Anyways, I have addressed your comments.

jesseduffield commented on pkg/commands/oscommands/os.go:
```diff
@@ -267,3 +268,16 @@ func GetLazygitPath() string {
+	}
+	title := fmt.Sprint(filepath.Base(path), " - Lazygit")
+	args := append([]string{"cmd", "/C", "title"}, strings.Split(title, " ")...)
+	return c.Cmd.NewShell(strings.Join(args, " ")).Run()
```
if I check the output in the Command Log panel I see `bash -c "cmd /C title lazygit - Lazygit"`. I'm on osx so it uses bash but I suspect on windows it shows `cmd /C "cmd /C title lazygit - Lazygit`. In effect, the NewShell function adds the `cmd` and `/C` arguments itself. So I'd see how you go removing those two args here

birgersp replied:
Thanks! I've removed the redundant args. Also I noticed that the OSCommand struct has os/platform information so I removed the dependency to the "runtime" package.

birgersp commented:
@jesseduffield you wrote

> if I check the output in the Command Log panel I see `bash -c "cmd /C title lazygit - Lazygit"`. I'm on osx so it uses bash (snip)

I wouldn't expect you to see any output in the Command Log panel from this, because of this condition:

```
	if c.Platform.OS != "windows" {
		return nil
	}
```

Is the `c.Platform.OS` property really "windows" in your debugging session (on OSX), too?

jesseduffield commented:
@birgersp sorry I didn't mention that I removed that check for the sake of seeing what would be logged

jesseduffield commented:
Code looks good. If you want to add a test feel free, but you'd need to store `getwd` on the OSCommand struct as a dependency so we can mock it out in the test. I'm not going to force you to go through that process if you don't want to! 

Running CI now

jesseduffield commented:
Merged. Thanks for your effort @birgersp !

birgersp commented:
Great! Thank you!

> Code looks good. If you want to add a test feel free, but you'd need to store `getwd` on the OSCommand struct as a dependency so we can mock it out in the test. I'm not going to force you to go through that process if you don't want to!
> 
> Running CI now

Would like to add a test, but I'm a complete beginner in Go so it's all a bit new for me. But I can try.
